libgalois/include/katana/PropertyGraph.h

```diff
@@ -842,6 +903,633 @@ CreateSymmetricGraph(PropertyGraph* pg);
+  }
+
+private:
+  // map an integer id to each unique edge label in the graph
+  using EdgeLabelToIndexMap = std::unordered_map<EdgeLabelTy, uint32_t>;
+  // each index refers to a unique edge label in the graph
+  using EdgeIndexToLabelMap = std::vector<EdgeLabelTy>;
+
+  EdgeLabelIndex(
+      EdgeLabelToIndexMap&& edge_label_to_index,
+      EdgeIndexToLabelMap&& edge_index_to_label, uint32_t num_labels) noexcept
+      : edge_label_to_index_map_(std::move(edge_label_to_index)),
+        edge_index_to_label_map_(std::move(edge_index_to_label)),
+        num_unique_edge_labels_(num_labels) {
+    KATANA_LOG_DEBUG_ASSERT(
+        num_labels == edge_index_to_label_map_.size() &&
+        num_labels == edge_label_to_index_map_.size());
+  }
+
+  EdgeLabelToIndexMap edge_label_to_index_map_;
+  EdgeIndexToLabelMap edge_index_to_label_map_;
+  uint32_t num_unique_edge_labels_ = 0u;
+};
+
+// TODO(amber): make OrigEdgeIDMap optional via template argument
+class KATANA_EXPORT EdgeShuffleTopology : public GraphTopology {
+  using Base = GraphTopology;
+  using OrigEdgeIDMap = katana::NUMAArray<Edge>;
+  using EdgeLabelTy = PropertyGraph::TypeSetID;
+
+public:
+  EdgeShuffleTopology() = default;
+  EdgeShuffleTopology(EdgeShuffleTopology&&) = default;
+  EdgeShuffleTopology& operator=(EdgeShuffleTopology&&) = default;
+
+  EdgeShuffleTopology(const EdgeShuffleTopology&) = delete;
+  EdgeShuffleTopology& operator=(const EdgeShuffleTopology&) = delete;
+
+  static EdgeShuffleTopology MakeTransposeCopy(const PropertyGraph* pg);
+  static EdgeShuffleTopology MakeOriginalCopy(const PropertyGraph* pg);
+
+  auto original_edge_id(Edge eid) const noexcept {
+    KATANA_LOG_DEBUG_ASSERT(eid < orig_edge_ids_.size());
+    return orig_edge_ids_[eid];
+  }
+
+  void SortEdgesByLabelThenDst() noexcept;
+
+protected:
+  NUMAArray<Edge>&& MoveOrigEdgeIDs() noexcept {
+    return std::move(orig_edge_ids_);
+  }
+
+  NUMAArray<Edge>& GetOrigEdgeIDs() noexcept { return orig_edge_ids_; }
+
+private:
+  // Needed for creating EdgeLabelAwareTopology from EdgeShuffleTopology
+  friend class EdgeLabelAwareTopology;
+
+  EdgeShuffleTopology(
+      const PropertyGraph* pg, NUMAArray<Edge>&& adj_indices,
+      NUMAArray<Node>&& dests, NUMAArray<Edge>&& orig_edge_ids) noexcept
+      : GraphTopology(std::move(adj_indices), std::move(dests)),
+        prop_graph_(pg),
+        orig_edge_ids_(std::move(orig_edge_ids)) {}
+
+  const PropertyGraph* prop_graph_;
+  OrigEdgeIDMap orig_edge_ids_;
+};
+
+/// store adjacency indices per each node such that they are divided by edge label type.
+/// Requires sorting the graph by edge label type
+class KATANA_EXPORT EdgeLabelAwareTopology : public GraphTopologyTypes {
+  // TODO(amber): add a template flag to choose between copying edge labels from
+  // PropertyGraph vs accessing the data in PropertyGraph through
+  // orig_edge_ids_
+
+protected:
+  using EdgeLabelTy = PropertyGraph::TypeSetID;
```

Comment:
This is part of the public API via `edge_begin` this should be public as well. Otherwise, callers cannot provide explicit types to the arguments to public API functions.

Reply:
done.